### PR TITLE
Add SyncBatchNorm

### PR DESCRIPTION
**Pull request by ceci3**
### PR types

Others

### PR changes

APIs

### Describe

Add SyncBatchNorm API for dygraph
1. because moving mean and moving variance computed all the time in the batch norm C++ kernel , so 'track_running_stats' is not kick in the API now. need to change batch norm kernel.
2. in the train mode, use mini-batch stats, in the eval mode, use global stats, not use `use_global_stas` and `trainable_stats` to jugde whether to use global or mini-batch stats

### CODE REVIEW

File: python/paddle/fluid/dygraph/nn.py [hunk 1]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+    :alias_main: paddle.nn.SyncBatchNorm
+	:alias: paddle.nn.SyncBatchNorm,paddle.nn.layer.SyncBatchNorm,paddle.nn.layer.norm.SyncBatchNorm
+	:old_api: paddle.fluid.dygraph.SyncBatchNorm
+
+    This interface is used to construct a callable object of the ``SyncBatchNorm`` class.
+    For more details, refer to code examples.
+    It implements the function of the Batch Normalization Layer and can be used 
+    as a normalizer function for conv2d and fully connected operations.
+    The data is normalized by the mean and variance of the channel based on all mini-batches
+    of the same process groups.
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `process group` is internal concept

File: python/paddle/fluid/dygraph/nn.py [hunk 2]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
 
 
+class SyncBatchNorm(layers.Layer):
+    """
+    :alias_main: paddle.nn.SyncBatchNorm
+	:alias: paddle.nn.SyncBatchNorm,paddle.nn.layer.SyncBatchNorm,paddle.nn.layer.norm.SyncBatchNorm
+	:old_api: paddle.fluid.dygraph.SyncBatchNorm
+
+    This interface is used to construct a callable object of the ``SyncBatchNorm`` class.
+    For more details, refer to code examples.
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `For more details, refer to code examples.` seems redundant

File: python/paddle/fluid/dygraph/nn.py [hunk 3]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+    It implements the function of the Batch Normalization Layer and can be used
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: reword this, emphasis synchronized, also usage not limited to conv or fc

File: python/paddle/fluid/dygraph/nn.py [hunk 4]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+    - :math:`\\gamma` : trainable proportional parameter
+    - :math:`\\beta` : trainable deviation parameter
+
+    **Note**:
+        moving mean and moving variance will be calculate whether `track_running_stats` is set to `True`
+        or `False`, we will fix it in the next time.
+
+    Parameters:
+        num_features(int): Indicate the number of channels of the input ``Tensor``.
+        eps(float, optional): The small value added to the variance to prevent division by zero. Default: 1e-5.
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: -> `epsilon`

File: python/paddle/fluid/dygraph/nn.py [hunk 5]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+        momentum(float, optional): The value used for the moving_mean and moving_var computation. Default: 0.9.
+        scale_attr(ParamAttr, optional): The parameter attribute for Parameter `scale`
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `scale_attr` and/or `bias_attr` can be set to `false`, in which case output is not scaled and/or shifted.

File: python/paddle/fluid/dygraph/nn.py [hunk 6]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+    Examples:
+        .. code-block:: python
+
+          import paddle.nn as nn
+          import paddle.fluid as fluid
+          from paddle.fluid.dygraph import to_variable
+          import numpy as np
+
+          x = np.random.random(size=(3, 10, 3, 7)).astype('float32')
+          with fluid.dygraph.guard():
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: use `paddle.enable_imperative()`, remove usage of `fluid` APIs

File: python/paddle/fluid/dygraph/nn.py [hunk 7]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+
+    Returns:
+        None
+
+    Examples:
+        .. code-block:: python
+
+          import paddle.nn as nn
+          import paddle.fluid as fluid
+          from paddle.fluid.dygraph import to_variable
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: paddle.to_variable

File: python/paddle/fluid/dygraph/nn.py [hunk 8]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+          import numpy as np
+
+          x = np.random.random(size=(3, 10, 3, 7)).astype('float32')
+          with fluid.dygraph.guard():
+              x = to_variable(x)
+              sync_batch_norm = nn.SyncBatchNorm(10)
+              hidden1 = sync_batch_norm(x)
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py:3266]: does this sample code actually work? without using `data_parallel`?

File: python/paddle/fluid/dygraph/nn.py [hunk 9]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+                 momentum=0.9,
+                 track_running_stats=True,
+                 scale_attr=None,
+                 bias_attr=None,
+                 name=None):
+        super(SyncBatchNorm, self).__init__()
+        self._scale_attr = scale_attr
+        self._bias_attr = bias_attr
+
+        assert bias_attr is not False, "bias_attr should not be False in batch_norm."
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: see comment above

File: python/paddle/fluid/dygraph/nn.py [hunk 10]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+
+        self._dtype = "float32"
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: remove this after the `dtype` change

File: python/paddle/fluid/dygraph/nn.py [hunk 11]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+
+        ### train mode: use mini-batch stats, eval mode: use global stats
+        if self.training:
+            use_global_stats = False
+            trainable_statistics = False
+        else:
+            use_global_stats = True
+            trainable_statistics = False
+
+        if in_dygraph_mode():
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py:3348]: refactor and extract the functional interface?

File: python/paddle/fluid/dygraph/nn.py [hunk 12]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+        self.weight.stop_gradient = self._scale_attr != None and self._scale_attr.learning_rate == 0.
+
+        self.bias = self.create_parameter(
+            attr=self._bias_attr,
+            shape=param_shape,
+            dtype=self._dtype,
+            is_bias=True)
+        self.bias.stop_gradient = self._scale_attr != None and self._scale_attr.learning_rate == 0.
+
+        self._mean = self.create_parameter(
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py:3320]: try implement `mean` and `var` as buffer if it works in static graph? see `Layer.register_buffer`

File: python/paddle/fluid/dygraph/nn.py [hunk 13]
```diff
@@ -3198,10 +3412,6 @@ class Flatten(layers.Layer):
         start_axis(int): first dim to flatten (default = 1)
         stop_axis(int): last dim to flatten (default = -1).
     
-    Returns:
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: belong to this PR?

File: python/paddle/fluid/dygraph/nn.py [hunk 14]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+
+    def __init__(self,
+                 num_features,
+                 eps=1e-05,
+                 momentum=0.9,
+                 track_running_stats=True,
+                 scale_attr=None,
+                 bias_attr=None,
+                 name=None):
+        super(SyncBatchNorm, self).__init__()
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py:3280]: does this fallback to regular batch norm or error out when not in data parallel mode?

File: python/paddle/fluid/dygraph/nn.py (hunk 1 above)
Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: done, thanks

File: python/paddle/fluid/dygraph/nn.py (hunk 2 above)
Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: deleted, thanks

File: python/paddle/fluid/dygraph/nn.py (hunk 3 above)
Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: done, thanks

File: python/paddle/fluid/dygraph/nn.py [hunk 15]
```diff
@@ -3184,6 +3184,220 @@ def forward(self, nodes_vector, edge_set):
...
+            dtype=self._dtype,
+            is_bias=True)
+        self.bias.stop_gradient = self._scale_attr != None and self._scale_attr.learning_rate == 0.
+
+        self._mean = self.create_parameter(
+            attr=ParamAttr(
+                name=None,
+                initializer=Constant(0.0),
+                trainable=False,
+                do_model_average=False),
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `do_model_average` should default to `True`, according to `efficientnet` and `efficientdet` implementation.

**Review by zhiqiu — COMMENTED** [review]
LGTM for `op_function_generator.cc`

File: python/paddle/fluid/dygraph/nn.py [hunk 16]
```diff
@@ -3184,6 +3184,214 @@ def forward(self, nodes_vector, edge_set):
...
+
+    When model in eval mode, the :math:`\\mu_{\\beta}`
+    and :math:`\\sigma_{\\beta}^{2}` are global or running statistics (moving_mean and moving_variance). 
+    It usually got from the pre-trained model. Calculated as follows:
+
+    .. math::
+        moving\_mean = moving\_mean * momentum + \mu_{\beta} * (1. - momentum) \quad &// global mean \\
+        moving\_variance = moving\_variance * momentum + \sigma_{\beta}^{2} * (1. - momentum) \quad &// global variance \\
+
+    The normalization function formula is as follows:
```
Comment (wanghaoshuang) [inline python/paddle/fluid/dygraph/nn.py]: The formula of normalization?

File: python/paddle/fluid/dygraph/nn.py [hunk 17]
```diff
@@ -3184,6 +3184,214 @@ def forward(self, nodes_vector, edge_set):
...
+        \\hat{x_i} &\\gets \\frac{x_i - \\mu_\\beta} {\\sqrt{\\
+        \\sigma_{\\beta}^{2} + \\eps}} \\qquad &//\ normalize \\\\
+        y_i &\\gets \\gamma \\hat{x_i} + \\beta \\qquad &//\ scale\ and\ shift
+
+    - :math:`\\eps` : add a smaller value to the variance to prevent division by zero
+    - :math:`\\gamma` : trainable proportional parameter
+    - :math:`\\beta` : trainable deviation parameter
+
+    **Note**:
+        moving mean and moving variance will be calculate whether `track_running_stats` is set to `True`
```
Comment (wanghaoshuang) [inline python/paddle/fluid/dygraph/nn.py]: calculated?

File: python/paddle/fluid/tests/unittests/parallel_dygraph_sync_batch_norm.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/parallel_dygraph_sync_batch_norm.py]: Copyright (c) 2020

File: python/paddle/fluid/tests/unittests/test_parallel_dygraph_sync_batch_norm.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wanghaoshuang) [inline python/paddle/fluid/tests/unittests/test_parallel_dygraph_sync_batch_norm.py]: 2020

Comment (ceci3) [reply python/paddle/fluid/tests/unittests/test_parallel_dygraph_sync_batch_norm.py]: done, thanks

File: python/paddle/fluid/tests/unittests/parallel_dygraph_sync_batch_norm.py (hunk 18 above)
Comment (ceci3) [reply python/paddle/fluid/tests/unittests/parallel_dygraph_sync_batch_norm.py]: done, thanks

File: python/paddle/fluid/dygraph/nn.py (hunk 17 above)
Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: done, thanks

File: python/paddle/fluid/dygraph/nn.py (hunk 16 above)
Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: done, thanks

File: python/paddle/fluid/dygraph/nn.py [hunk 20]
```diff
@@ -3184,6 +3184,214 @@ def forward(self, nodes_vector, edge_set):
...
+          import paddle.nn as nn
+          import paddle.fluid as fluid
+          from paddle.fluid.dygraph import to_variable
+          import numpy as np
+
+          x = np.random.random(size=(3, 10, 3, 7)).astype('float32')
+          with fluid.dygraph.guard():
+              x = to_variable(x)
+              if fluid.is_compiled_with_cuda():
+                  sync_batch_norm = nn.SyncBatchNorm(10)
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py]: 辛苦用注释的形式给出输入、输出示例。如果PR着急，辛苦merge后再提一个PR进行更新。

File: python/paddle/fluid/dygraph/nn.py [hunk 21]
```diff
@@ -3184,6 +3184,214 @@ def forward(self, nodes_vector, edge_set):
...
+                  hidden1 = sync_batch_norm(x)
+    """
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py:3269]: 辛苦给出文档的预览图，防止文档格式解析混乱问题

**Review by zhiqiu — DISMISSED** [review]
LGTM for op_function_generator.cc

**Comment by ceci3** [comment]
image

File: python/paddle/fluid/dygraph/nn.py [hunk 22]
```diff
@@ -3184,6 +3184,216 @@ def forward(self, nodes_vector, edge_set):
         return self._helper.append_activation(pre_activation, act=self._act)
 
 
+class SyncBatchNorm(layers.Layer):
+    """
+    :alias_main: paddle.nn.SyncBatchNorm
+	:alias: paddle.nn.SyncBatchNorm,paddle.nn.layer.SyncBatchNorm,paddle.nn.layer.norm.SyncBatchNorm
+	:old_api: paddle.fluid.dygraph.SyncBatchNorm
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py]: 最新规定：alias部分可以全部删除哈，辛苦删除

File: python/paddle/fluid/dygraph/nn.py [hunk 23]
```diff
@@ -3184,6 +3184,216 @@ def forward(self, nodes_vector, edge_set):
...
+    Examples:
+        .. code-block:: python
+
+          import paddle.nn as nn
+          import paddle.fluid as fluid
+          from paddle.fluid.dygraph import to_variable
+          import numpy as np
+
+          x = np.array([[[[0.3, 0.4], [0.3, 0.07]], [[0.83, 0.37], [0.18, 0.93]]]]).astype('float32')
+          with fluid.dygraph.guard():
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py]: 示例可以直接用paddle2.0-alpha的api
with paddle.imperative.guard():

File: python/paddle/fluid/dygraph/nn.py [hunk 24]
```diff
@@ -3184,6 +3184,216 @@ def forward(self, nodes_vector, edge_set):
...
+
+    Returns:
+        None
+
+    Examples:
+        .. code-block:: python
+
+          import paddle.nn as nn
+          import paddle.fluid as fluid
+          from paddle.fluid.dygraph import to_variable
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py]: 辛苦用2.0的api
paddle.imperative.to_variable

File: python/paddle/fluid/dygraph/nn.py [hunk 25]
```diff
@@ -3184,6 +3184,216 @@ def forward(self, nodes_vector, edge_set):
...
+          import numpy as np
+
+          x = np.array([[[[0.3, 0.4], [0.3, 0.07]], [[0.83, 0.37], [0.18, 0.93]]]]).astype('float32')
+          with fluid.dygraph.guard():
+              x = to_variable(x)
+              if fluid.is_compiled_with_cuda():
```
Comment (swtkiwi) [inline python/paddle/fluid/dygraph/nn.py]: 需要用2.0的api，不再展示fluid

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/dygraph/nn.py [hunk 26]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+        \ mini-batch\ mean \\\\
+        \\sigma_{\\beta}^{2} &\\gets \\frac{1}{m} \\sum_{i=1}^{m}(x_i - \\
+        \\mu_{\\beta})^2 \\qquad &//\ mini-batch\ variance \\\\
+
+    - :math:`x` : mini-batch data
+    - :math:`m` : the size of the mini-batch data
+
+    When model in eval mode, the :math:`\\mu_{\\beta}`
+    and :math:`\\sigma_{\\beta}^{2}` are global or running statistics (moving_mean and moving_variance). 
+    It usually got from the pre-trained model. Calculated as follows:
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: reword

File: python/paddle/fluid/dygraph/nn.py [hunk 27]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+        \\mu_{\\beta} &\\gets \\frac{1}{m} \\sum_{i=1}^{m} x_i \\qquad &//\\
+        \ mini-batch\ mean \\\\
+        \\sigma_{\\beta}^{2} &\\gets \\frac{1}{m} \\sum_{i=1}^{m}(x_i - \\
+        \\mu_{\\beta})^2 \\qquad &//\ mini-batch\ variance \\\\
+
+    - :math:`x` : mini-batch data
+    - :math:`m` : the size of the mini-batch data
+
+    When model in eval mode, the :math:`\\mu_{\\beta}`
+    and :math:`\\sigma_{\\beta}^{2}` are global or running statistics (moving_mean and moving_variance).
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: "global or running statistics", depend on what?  too confusing, need more explanation.

File: python/paddle/fluid/dygraph/nn.py [hunk 28]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+    The formula of normalization is as follows:
+ 
+    ..  math::
+
+        \\hat{x_i} &\\gets \\frac{x_i - \\mu_\\beta} {\\sqrt{\\
+        \\sigma_{\\beta}^{2} + \\eps}} \\qquad &//\ normalize \\\\
+        y_i &\\gets \\gamma \\hat{x_i} + \\beta \\qquad &//\ scale\ and\ shift
+
+    - :math:`\\eps` : add a smaller value to the variance to prevent division by zero
+    - :math:`\\gamma` : trainable proportional parameter
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: "proportional" and "deviation"  sounds weird.

File: python/paddle/fluid/dygraph/nn.py [hunk 29]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+    - :math:`\\beta` : trainable deviation parameter
+
+    **Note**:
+        moving mean and moving variance will be calculated whether `track_running_stats` is set to `True`
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: just remove it for now, add back later when the underlying implementation is in place.

File: python/paddle/fluid/dygraph/nn.py [hunk 30]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+        self._bias_attr = bias_attr
+        self._num_features = num_features
+        self._data_layout = data_format
+        self._momentum = momentum
+        self._epsilon = epsilon
+        self._track_running_stats = track_running_stats
+
+        param_shape = [self._num_features]
+
+        ### TODO(lvmengsi): remove create param when weight_attr=False in python when BatchNorm kernel support
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `weight` and `bias` should be controlled separately. also, describe the behavior in docstring

File: python/paddle/fluid/dygraph/nn.py [hunk 31]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+            attr=ParamAttr(
+                name=None,
+                initializer=Constant(1.0),
+                trainable=False,
+                do_model_average=True),
+            shape=param_shape,
+            dtype=self._dtype)
+        self._variance.stop_gradient = True
+
+    def forward(self, input):
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: `input` -> `x`

File: python/paddle/fluid/dygraph/nn.py [hunk 32]
```diff
@@ -3185,6 +3186,262 @@ def forward(self, nodes_vector, edge_set):
...
+    #        if layer._weight_attr != False and layer._bias_attr != False:
+    #            with no_grad():
+    #                layer_output.weight = layer.weight
+    #                layer_output.bias = layer.bias
+    #        layer_output._mean = layer._mean
+    #        layer_output._variance = layer._variance
+
+    #    for name, sublayer in layer.named_sublayer():
+    #        layer_output.add_sublayer(name, cls.convert_sync_batchnorm(sublayer))
+    #    del layer
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: what does this do? what happens if it is not deleted?

**Review by zhiqiu — DISMISSED** [review]
LGTM for op_function_generator.cc

File: python/paddle/fluid/dygraph/nn.py [hunk 33]
```diff
@@ -20,7 +20,6 @@
 from ..layers import nn as F
 from .. import dygraph_utils
 from . import layers
-from .base import no_grad
```
Comment (willthefrog) [inline python/paddle/fluid/dygraph/nn.py]: seems unrelated to current PR?

Comment (ceci3) [reply python/paddle/fluid/dygraph/nn.py]: deleted, thanks

**Review by willthefrog — DISMISSED** [review]
need subsequent change for converting batchnorms, otherwise LGTM

**Review by willthefrog — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM
